src/allmydata/mutable/publish.py

```diff
@@ -955,12 +960,31 @@ def update_goal(self):
             old_assignments.add(server, shnum)
 
         serverlist = []
-        for i, server in enumerate(self.full_serverlist):
-            serverid = server.get_serverid()
-            if server in self.bad_servers:
-                continue
-            entry = (len(old_assignments.get(server, [])), i, serverid, server)
-            serverlist.append(entry)
+
+        action = start_action(
+            action_type=u"mutable:upload:update_goal",
+            homeless_shares=len(homeless_shares),
+        )
+        with action:
+            for i, server in enumerate(self.full_serverlist):
+                serverid = server.get_serverid()
+                if server in self.bad_servers:
+                    Message.log(
+                        message_type=u"mutable:upload:bad-server",
+                        server_id=serverid,
+                    )
+                    continue
+                # if we have >= 1 grid-managers, this checks that we have
+                # a valid certificate for this server
+                if not server.upload_permitted():
+                    Message.log(
+                        message_type=u"mutable:upload:no-gm-certs",
+                        server_id=serverid,
+                    )
+                    continue
+
```

Comment:
Maybe.

I was going for "minimal changes" here (it looks like more because eliot forces you to re-indent to add logging). Definitely "mutables" could use some re-factoring, but ...